[IMP] mail: share screen audio

Atovange opened the pull request:
With this commit it will be possible to also share the audio when sharing the screen.

task-4450111

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/194108.png)](https://mergebot.odoo.com/odoo/odoo/pull/194108)

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/call/common/rtc_service.js:
```diff
@@ -1356,7 +1365,19 @@ export class Rtc extends Record {
             this.linkVoiceActivationDebounce();
-            await this.network?.updateUpload("audio", this.state.audioTrack);
+            if (this.state.screenAudioTrack) {
+                const audioContext = new AudioContext();
```
At some point the `AudioContext` should be [closed](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/close), when a new one is made, and in `clear()`

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/call/common/rtc_service.js:
```diff
@@ -1356,7 +1365,19 @@ export class Rtc extends Record {
+            if (this.state.screenAudioTrack) {
+                const audioContext = new AudioContext();
+                const micSource = audioContext.createMediaStreamSource(audioStream);
+                const screenAudioStream = new MediaStream([this.state.screenAudioTrack]);
+                const screenAudioSource = audioContext.createMediaStreamSource(screenAudioStream);
```
Have you tried [`createMediaStreamTrackSource()`](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/createMediaStreamTrackSource)? It would remove the need for a new `MediaStream` and the `let audioStream` above.

Atovange replied:
Unfortunately it has very bad browser support: https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/createMediaStreamTrackSource#browser_compatibility

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/call/common/rtc_service.js:
```diff
@@ -1305,11 +1309,15 @@ export class Rtc extends Record {
             }
         }
+        if (this.state.screenAudioTrack) {
+            this.resetAudioTrack({ force: true });
```
I wouldn't call `resetAudioTrack()` here, it's doing a lot of unnecessary things, I think that there should be a new function that takes care of merging `this.state.screenAudioTrack` and `this.state.audioTrack` and updating the network, called both here and in`resetAudioTrack()`.

ThanhDodeurOdoo reviewed (commented):
I haven't tested yet, but it looks good

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/call/common/rtc_service.js:
```diff
@@ -1356,7 +1365,19 @@ export class Rtc extends Record {
+                const mixedAudioDestination = audioContext.createMediaStreamDestination();
+                micSource.connect(mixedAudioDestination);
+                screenAudioSource.connect(mixedAudioDestination);
+                const mixedAudioStreamTrack = mixedAudioDestination.stream.getAudioTracks()[0];
```
This should stored in the state, it will be needed if the network changes,

For example when we upgrade from P2P to SFU, or fallback from SFU to P2P, we call `updateUpload()` on the new network to pass all the tracks that already exist.

Most (if not all) `updateUpload("audio", this.state.audioTrack)` should be replaced by `updateUpload("audio", this.state.mixedAudioStreamTrack)`, we could think of `mixedAudioStreamTrack` as an "output audio track" that sometimes is a mix, sometimes is just `audioTrack` and sometimes even just `screenAudioTrack`

Atovange commented:
Summary diff for the reviewer:

- I made `audioTrack` the main audio track to pass to `updateUpload` and introduced `micAudioTrack` to handle specifically the microphone audio.
- I extracted the mixing logic to `updateAudioTrack` that will update the `audioTrack` in the state based on the available audio tracks (if any).

alexkuhn reviewed (approved):
LGTM

Some things that could be done in follow-up PRs:
- rename functions "audio" that are mic-specific, e.g. `refreshAudioStatus` looks mic-specific
- some features could be added with screen-sharing audio, e.g. the context menu can have 2 volume gauges, one for mic and one for screen-sharing.

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/call/common/rtc_service.js:
```diff
@@ -1305,31 +1314,61 @@ export class Rtc extends Record {
+    async resetAudioTrack({ force = false }) {
+        this.state.micAudioTrack?.stop();
+        this.state.micAudioTrack = undefined;
+        this.state.screenAudioTrack?.stop();
+        this.state.screenAudioTrack = undefined;
```
`resetAudioTrack()` is only for the microphone (maybe it should be renamed), if you stop the `screenAudioTrack` here, screen audio will be tied to the lifespan of the microphone audio (which I don't think is expected).

For example, if you change the setting of your audio input device, `resetAudioTrack()` will be called to get a new one from the right device, but there it will also stop the screen audio.

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/call/common/rtc_service.js:
```diff
@@ -1038,8 +1040,11 @@ export class Rtc extends Record {
         browser.clearTimeout(this.sfuTimeout);
         this.sfuClient = undefined;
         this.network = undefined;
+        this.audioContext?.close();
```
When an `AudioContext` is closed, it will throw an error (`e.name === "InvalidStateError"`) if it's already closed, you could check the state before, or try/catch or set it as undefined when it's closed.

If you don't set it as undefined when it's closed, leaving a call and entering one again will lead to a traceback whenever it's reached (as we would call `close()` on the closed audioContext of the previous call).

also, maybe a bit of nitpicking but I would declare the audioContext and document it in the class property declarations

ThanhDodeurOdoo reviewed (commented):
2 last things and then I think it's good

@alexkuhn `some features could be added with screen-sharing audio, e.g. the context menu can have 2 volume gauges, one for mic and one for screen-sharing`

Note (for when we need that feature) that the approach would need to change to make it possible, currently the two tracks are mixed, not multiplexed, so on the receiving end we only get 1 track that cannot be demuxed. For individually controlled volumes, it would need two individual audio tracks (and allowing it SFU-side too).
